pkg/api/validation/validation.go

```diff
@@ -555,15 +555,80 @@ func validateEnv(vars []api.EnvVar) errs.ValidationErrorList {
 		vErrs := errs.ValidationErrorList{}
 		if len(ev.Name) == 0 {
 			vErrs = append(vErrs, errs.NewFieldRequired("name"))
-		}
-		if !util.IsCIdentifier(ev.Name) {
+		} else if !util.IsCIdentifier(ev.Name) {
 			vErrs = append(vErrs, errs.NewFieldInvalid("name", ev.Name, cIdentifierErrorMsg))
 		}
+		vErrs = append(vErrs, validateEnvVarSources(ev)...)
 		allErrs = append(allErrs, vErrs.PrefixIndex(i)...)
 	}
 	return allErrs
 }
 
+func validateEnvVarSources(ev api.EnvVar) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	if ev.EnvVarSource == nil {
+		return allErrs
+	}
+
+	numSources := 0
+
+	switch {
+	case ev.ObjectInfo != nil:
+		numSources++
+		allErrs = append(allErrs, validateObjectInfo(ev.ObjectInfo)...)
+	case ev.HostInfo != nil:
+		numSources++
+		allErrs = append(allErrs, validateHostInfo(ev.HostInfo)...)
+	}
+
+	if ev.Value != "" && numSources >= 1 {
+		allErrs = append(allErrs, errs.NewFieldInvalid("", "", "sources cannot be specified when value is set"))
```

Comment:
error message could be more helpful.  "non-literal source may not be specified..." ?  Something a bit clearer.